src/plugins/workspace/server/permission_control/client.ts

```diff
@@ -48,11 +53,46 @@ export class SavedObjectsPermissionControl {
     this.logger = logger;
   }
 
+  private generateSavedObjectKey = ({ type, id }: { type: string; id: string }) => {
+    return `${type}${id}`;
+  };
+
   private async bulkGetSavedObjects(
     request: OpenSearchDashboardsRequest,
     savedObjects: SavedObjectsBulkGetObject[]
   ) {
-    return (await this.getScopedClient?.(request)?.bulkGet(savedObjects))?.saved_objects || [];
+    const requestKey = ensureRawRequest(request);
+    const savedObjectsToGet = savedObjects.filter(
+      (savedObject) =>
+        !this._savedObjectCache.get(requestKey)?.[this.generateSavedObjectKey(savedObject)]
+    );
+    const retrievedSavedObjects =
+      savedObjectsToGet.length > 0
+        ? (await this.getScopedClient?.(request)?.bulkGet(savedObjectsToGet))?.saved_objects || []
+        : [];
+
+    const retrievedSavedObjectsMap: { [key: string]: SavedObject } = {};
+    retrievedSavedObjects.forEach((savedObject) => {
+      const savedObjectKey = this.generateSavedObjectKey(savedObject);
+      if (this._shouldCachedSavedObjects.get(requestKey)?.includes(savedObjectKey)) {
```

Comment:
Seems we can detect if there is cache already by looking up directly in the `_savedObjectCache` and making it the source of truth.